File: packages/scheduling-engine/src/analysis/gen-decisions.ts
// ═══════════════════════════════════════════════════════════
//  INCOMPOL PLAN — Gen Decisions
//  Generates UI-facing replan proposals from blocked operations.
//  Scores alternative machines using running capacity tracking.
//  Migrated from NikufraEngine.tsx genDecisions()
// ═══════════════════════════════════════════════════════════

import { DAY_CAP } from '../constants.js';
import type { Block, MoveAction } from '../types/blocks.js';
import type { EMachine, EOp, ETool } from '../types/engine.js';
import { capAnalysis } from './cap-analysis.js';

// ── Types ────────────────────────────────────────────────

export type DecisionSeverity = 'critical' | 'high' | 'medium' | 'low';
export type DecisionKind = 'replan' | 'blocked';

export interface ReplanProposal {
  id: string;
  opId: string;
  type: DecisionKind;
  severity: DecisionSeverity;
  title: string;
  desc: string;
  reasoning: string[];
  impact: Record<string, unknown> | null;
  action: MoveAction | null;
}

// ── Main function ────────────────────────────────────────

/**
 * Generate UI-facing replan proposals for blocked operations.
 *
 * For each blocked operation (machine down), evaluates whether an
 * alternative machine is available, scores candidates using a running
 * capacity tracker (so each subsequent proposal accounts for cumulative
 * load from prior proposals), and produces a sorted list of proposals.
 *
 * Also emits warnings for operations below economic lot size.
 */
export function genDecisions(
  ops: EOp[],
  mSt: Record<string, string>,
  _tSt: Record<string, string>,
  moves: MoveAction[],
  blocks: Block[],
  machines: EMachine[],
  TM: Record<string, ETool>,
  _focusIds: string[],
  tools: ETool[],
): ReplanProposal[] {
  const decs: ReplanProposal[] = [];
  const cap = capAnalysis(blocks, machines);

  // Running capacity tracker — accumulates load from proposed moves
  // so each subsequent decision sees the combined load of all prior decisions
  const runCap: Record<string, { prod: number; setup: number }[]> = {};
  for (const [mId, days] of Object.entries(cap)) {
    runCap[mId] = days.map((d) => ({ prod: d.prod, setup: d.setup }));
  }

  const blkOps = new Map<string, Block>();
  blocks
    .filter((b) => b.type === 'blocked' && !moves.find((mv) => mv.opId === b.opId))
    .forEach((b) => {
      if (!blkOps.has(b.opId)) blkOps.set(b.opId, b);
    });

  // Sort by severity: highest priority first (stock-zero + high-backlog first)
  const sortedBlk = [...blkOps.values()].sort((a, b) => {
    const toolA = TM[a.toolId],
      toolB = TM[b.toolId];
    const opA = ops.find((o) => o.id === a.opId),
      opB = ops.find((o) => o.id === b.opId);
    const sevA = (toolA && toolA.stk === 0 ? 10 : 0) + (opA ? opA.atr : 0);
    const sevB = (toolB && toolB.stk === 0 ? 10 : 0) + (opB ? opB.atr : 0);
    return sevB - sevA;
  });

  sortedBlk.forEach((b) => {
    const tool = TM[b.toolId],
      op = ops.find((o) => o.id === b.opId);
    if (!tool || !op) return;
    if (tool.pH <= 0) return; // Guard: skip operations with rate=0
    const totalPcs = op.d.reduce((a, v) => a + Math.max(v, 0), 0) + Math.max(op.atr, 0);
    const totalH = totalPcs / tool.pH,
      setupMinVal = tool.sH * 60;
    const hasStk = tool.stk > 0,
      stkDays = hasStk ? tool.stk / (tool.pH * 16) : 0;
    const sev =
      (op.atr > 20000 ? 4 : op.atr > 5000 ? 3 : op.atr > 0 ? 2 : 0) +
      (tool.stk === 0 && tool.lt > 0 ? 3 : 0) +
      (totalPcs > 20000 ? 2 : totalPcs > 5000 ? 1 : 0);
    const severity: DecisionSeverity =
      sev >= 5 ? 'critical' : sev >= 3 ? 'high' : sev >= 1 ? 'medium' : 'low';
    const R: string[] = [];

    if (b.reason === 'tool_down') {
      R.push(`Ferramenta ${b.toolId} AVARIADA.`);
      if (op.atr > 0) R.push(`Backlog: ${op.atr.toLocaleString()} pcs.`);
      decs.push({
        id: `D_${b.opId}_TF`,
        opId: b.opId,
        type: 'blocked',
        severity: op.atr > 0 ? 'critical' : 'high',
        title: `${b.toolId} avariada`,
        desc: `${b.nm} (${b.sku})`,
        reasoning: R,
        impact: { pcsLost: totalPcs, hrsLost: totalH.toFixed(1) },
        action: null,
      });
      return;
    }

    R.push(`Máquina ${b.origM} DOWN → ${b.toolId}/${b.sku} afetada.`);
    R.push(`Volume: ${totalPcs.toLocaleString()} pcs (${totalH.toFixed(1)}h).`);

    if (!b.hasAlt) {
      R.push('Sem alternativa (ISOP).');
      R.push(
        !hasStk
          ? 'STOCK ZERO → paragem.'
          : `Buffer: ${tool.stk.toLocaleString()} pcs (≈${stkDays.toFixed(1)}d).`,
      );
      decs.push({
        id: `D_${b.opId}_NA`,
        opId: b.opId,
        type: 'blocked',
        severity: !hasStk ? 'critical' : severity,
        title: `${b.toolId} sem alternativa`,
        desc: b.nm,
        reasoning: R,
        impact: { pcsLost: totalPcs, hrsLost: totalH.toFixed(1), stkDays: stkDays.toFixed(1) },
        action: null,
      });
      return;
    }

    const candidates: string[] = [];
    if (b.altM && mSt[b.altM] !== 'down') candidates.push(b.altM);

    if (candidates.length === 0) {
      R.push(`Alt. ${b.altM} TAMBÉM DOWN.`);
      decs.push({
        id: `D_${b.opId}_AD`,
        opId: b.opId,
        type: 'blocked',
        severity: 'critical',
        title: `${b.toolId}: ambas DOWN`,
        desc: b.nm,
        reasoning: R,
        impact: { pcsLost: totalPcs, hrsLost: totalH.toFixed(1) },
        action: null,
      });
      return;
    }

    // Score candidates using RUNNING capacity (includes load from prior decisions)
    const gNDays = op.d.length || 8;
    const scored = candidates.map((cId) => {
      const dLoad = Array.from({ length: gNDays }, (_, di) => {
        const dc = runCap[cId]?.[di] || { prod: 0, setup: 0 };
        const addProd = op.d[di] > 0 ? (op.d[di] / tool.pH) * 60 : 0;
        const firstProdDay = op.d.findIndex((v) => v > 0);
        const addSetup = di === firstProdDay || di === 0 ? setupMinVal : 0;
        const total = dc.prod + dc.setup + addProd + addSetup;
        return {
          day: di,
          current: dc.prod + dc.setup,
          added: addProd + addSetup,
          total,
          util: total / DAY_CAP,
        };
      });
      const peak = Math.max(...dLoad.map((d) => d.util));
      const overDays = dLoad.filter((d) => d.util > 1.0);
      const sharedMP =
        tool.mp != null &&
        tools.filter(
          (t2) => t2.mp === tool.mp && t2.id !== tool.id && (t2.m === cId || t2.alt === cId),
        ).length > 0;
      return {
        mId: cId,
        dLoad,
        peak,
        overDays,
        sharedMP,
        score: peak * 100 + overDays.length * 50 + setupMinVal * 0.1 - (sharedMP ? 30 : 0),
      };
    });
    scored.sort((a, b) => a.score - b.score);
    const best = scored[0];

    // Update running capacity with proposed move load
    if (runCap[best.mId]) {
      for (let di = 0; di < gNDays; di++) {
        if (!runCap[best.mId][di]) runCap[best.mId][di] = { prod: 0, setup: 0 };
        const addProd = op.d[di] > 0 ? (op.d[di] / tool.pH) * 60 : 0;
        const firstProdDay = op.d.findIndex((v) => v > 0);
        const addSetup = di === firstProdDay || di === 0 ? setupMinVal : 0;
        runCap[best.mId][di].prod += addProd;
        runCap[best.mId][di].setup += addSetup;
      }
    }

    R.push(`Alt. disponível: ${best.mId}.`);
    if (best.overDays.length > 0)
      R.push(
        `${best.mId} sobrecarga ${best.overDays.length}d (inclui ${decs.filter((d) => d.action?.toM === best.mId).length} ops já propostas).`,
      );
    else R.push(`Capacidade ${best.mId}: pico ${(best.peak * 100).toFixed(0)}% — OK.`);
    R.push(`Setup: +${setupMinVal}min (${tool.sH}h).`);
    if (best.sharedMP) R.push(`MP ${tool.mp} partilhada — agrupar.`);
    if (tool.op > 1) R.push(`Requer ${tool.op} operadores.`);
    if (!hasStk && tool.lt > 0) R.push('STOCK ZERO — OTD em risco.');
    R.push(`→ Mover ${b.toolId} → ${best.mId}.`);

    decs.push({
      id: `D_${b.opId}_RP`,
      opId: b.opId,
      type: 'replan',
      severity,
      title: `${b.toolId} → ${best.mId}`,
      desc: `${b.nm} (${b.sku})`,
      reasoning: R,
      impact: {
        fromM: b.origM,
        toM: best.mId,
        setupMin: setupMinVal,
        pcs: totalPcs,
        hrs: totalH.toFixed(1),
        destPeak: (best.peak * 100).toFixed(0),
        overDays: best.overDays.length,
        ops: tool.op,
        stockRisk: !hasStk && tool.lt > 0,
        atr: op.atr,
        sharedMP: best.sharedMP,
        dLoad: best.dLoad,
      },
      action: { opId: b.opId, toM: best.mId },
    });
  });

  // Lote económico warnings
  blocks
    .filter((b) => b.belowMinBatch && b.type === 'ok')
    .forEach((b) => {
      const tool = TM[b.toolId];
      if (!tool) return;
      decs.push({
        id: `D_${b.opId}_LT`,
        opId: b.opId,
        type: 'replan',
        severity: 'low',
        title: `${b.sku} abaixo lote econ.`,
        desc: `${b.qty.toLocaleString()} < ${tool.lt.toLocaleString()} pcs`,
        reasoning: [
          `Qty ${b.qty.toLocaleString()} abaixo lote económico ${tool.lt.toLocaleString()}.`,
          `Considerar agrupar com próxima encomenda.`,
        ],
        impact: { qty: b.qty, lotEconomic: tool.lt },
        action: null,
      });
    });

  const sO: Record<string, number> = { critical: 0, high: 1, medium: 2, low: 3 };
  decs.sort((a, b) =>
    a.type === 'replan' && b.type !== 'replan'
      ? -1
      : b.type === 'replan' && a.type !== 'replan'
        ? 1
        : (sO[a.severity] ?? 3) - (sO[b.severity] ?? 3),
  );
  return decs;
}

File: apps/backend/src/domain/scheduling/analysis/gen_decisions.py
"""Gen Decisions — port of analysis/gen-decisions.ts.

Generates UI-facing replan proposals from blocked operations.
Scores alternative machines using running capacity tracking.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Literal

from ..constants import DAY_CAP
from ..types import Block, EMachine, EOp, ETool, MoveAction
from .cap_analysis import cap_analysis

DecisionSeverity = Literal["critical", "high", "medium", "low"]
DecisionKind = Literal["replan", "blocked"]


@dataclass
class ReplanProposal:
    id: str
    op_id: str
    type: str  # DecisionKind
    severity: str  # DecisionSeverity
    title: str
    desc: str
    reasoning: list[str] = field(default_factory=list)
    impact: dict[str, Any] | None = None
    action: dict[str, str] | None = None


def _fmt_num(n: int) -> str:
    return f"{n:,}"


def gen_decisions(
    ops: list[EOp],
    m_st: dict[str, str],
    t_st: dict[str, str],
    moves: list[MoveAction],
    blocks: list[Block],
    machines: list[EMachine],
    tool_map: dict[str, ETool],
    focus_ids: list[str],
    tools: list[ETool],
) -> list[ReplanProposal]:
    """Generate UI-facing replan proposals for blocked operations."""
    decs: list[ReplanProposal] = []
    cap = cap_analysis(blocks, machines)

    # Running capacity tracker
    run_cap: dict[str, list[dict[str, float]]] = {}
    for m_id, days in cap.items():
        run_cap[m_id] = [{"prod": d.get("prod", 0), "setup": d.get("setup", 0)} for d in days]

    # Collect blocked ops not yet moved
    move_op_ids = {mv.op_id for mv in moves}
    blk_ops: dict[str, Block] = {}
    for b in blocks:
        if b.type == "blocked" and b.op_id not in move_op_ids:
            if b.op_id not in blk_ops:
                blk_ops[b.op_id] = b

    # Sort by severity: stock-zero + high-backlog first
    def _sort_key(b: Block) -> float:
        tool = tool_map.get(b.tool_id)
        op = next((o for o in ops if o.id == b.op_id), None)
        sev = (10 if tool and tool.stk == 0 else 0) + (op.atr if op else 0)
        return -sev  # descending

    sorted_blk = sorted(blk_ops.values(), key=_sort_key)

    for b in sorted_blk:
        tool = tool_map.get(b.tool_id)
        op = next((o for o in ops if o.id == b.op_id), None)
        if not tool or not op:
            continue
        if tool.pH <= 0:
            continue

        total_pcs = sum(max(v, 0) for v in op.d) + max(op.atr, 0)
        total_h = total_pcs / tool.pH
        setup_min_val = tool.sH * 60
        has_stk = tool.stk > 0
        stk_days = tool.stk / (tool.pH * 16) if has_stk else 0

        sev_score = (
            (4 if op.atr > 20000 else 3 if op.atr > 5000 else 2 if op.atr > 0 else 0)
            + (3 if tool.stk == 0 and tool.lt > 0 else 0)
            + (2 if total_pcs > 20000 else 1 if total_pcs > 5000 else 0)
        )
        severity: str = (
            "critical"
            if sev_score >= 5
            else "high"
            if sev_score >= 3
            else "medium"
            if sev_score >= 1
            else "low"
        )
        reasoning: list[str] = []

        if b.reason == "tool_down":
            reasoning.append(f"Ferramenta {b.tool_id} AVARIADA.")
            if op.atr > 0:
                reasoning.append(f"Backlog: {_fmt_num(op.atr)} pcs.")
            decs.append(
                ReplanProposal(
                    id=f"D_{b.op_id}_TF",
                    op_id=b.op_id,
                    type="blocked",
                    severity="critical" if op.atr > 0 else "high",
                    title=f"{b.tool_id} avariada",
                    desc=f"{b.nm} ({b.sku})",
                    reasoning=reasoning,
                    impact={"pcsLost": total_pcs, "hrsLost": f"{total_h:.1f}"},
                    action=None,
                )
            )
            continue

        reasoning.append(f"Máquina {b.orig_m} DOWN → {b.tool_id}/{b.sku} afetada.")
        reasoning.append(f"Volume: {_fmt_num(total_pcs)} pcs ({total_h:.1f}h).")

        if not b.has_alt:
            reasoning.append("Sem alternativa (ISOP).")
            if not has_stk:
                reasoning.append("STOCK ZERO → paragem.")
            else:
                reasoning.append(f"Buffer: {_fmt_num(tool.stk)} pcs (≈{stk_days:.1f}d).")
            decs.append(
                ReplanProposal(
                    id=f"D_{b.op_id}_NA",
                    op_id=b.op_id,
                    type="blocked",
                    severity="critical" if not has_stk else severity,
                    title=f"{b.tool_id} sem alternativa",
                    desc=b.nm,
                    reasoning=reasoning,
                    impact={
                        "pcsLost": total_pcs,
                        "hrsLost": f"{total_h:.1f}",
                        "stkDays": f"{stk_days:.1f}",
                    },
                    action=None,
                )
            )
            continue

        candidates: list[str] = []
        if b.alt_m and m_st.get(b.alt_m) != "down":
            candidates.append(b.alt_m)

        if not candidates:
            reasoning.append(f"Alt. {b.alt_m} TAMBÉM DOWN.")
            decs.append(
                ReplanProposal(
                    id=f"D_{b.op_id}_AD",
                    op_id=b.op_id,
                    type="blocked",
                    severity="critical",
                    title=f"{b.tool_id}: ambas DOWN",
                    desc=b.nm,
                    reasoning=reasoning,
                    impact={"pcsLost": total_pcs, "hrsLost": f"{total_h:.1f}"},
                    action=None,
                )
            )
            continue

        # Score candidates using RUNNING capacity
        g_n_days = len(op.d) if op.d else 8
        scored: list[dict] = []
        for c_id in candidates:
            d_load = []
            for di in range(g_n_days):
                dc = (
                    run_cap[c_id][di]
                    if c_id in run_cap and di < len(run_cap[c_id])
                    else {"prod": 0, "setup": 0}
                )
                add_prod = (op.d[di] / tool.pH) * 60 if di < len(op.d) and op.d[di] > 0 else 0
                first_prod_day = next((i for i, v in enumerate(op.d) if v > 0), 0)
                add_setup = setup_min_val if di == first_prod_day or di == 0 else 0
                total = dc["prod"] + dc["setup"] + add_prod + add_setup
                d_load.append(
                    {
                        "day": di,
                        "current": dc["prod"] + dc["setup"],
                        "added": add_prod + add_setup,
                        "total": total,
                        "util": total / DAY_CAP if DAY_CAP > 0 else 0,
                    }
                )

            peak = max((dl["util"] for dl in d_load), default=0)
            over_days = [dl for dl in d_load if dl["util"] > 1.0]
            shared_mp = tool.mp is not None and any(
                t2.mp == tool.mp and t2.id != tool.id and (t2.m == c_id or t2.alt == c_id)
                for t2 in tools
            )
            score = (
                peak * 100 + len(over_days) * 50 + setup_min_val * 0.1 - (30 if shared_mp else 0)
            )
            scored.append(
                {
                    "mId": c_id,
                    "dLoad": d_load,
                    "peak": peak,
                    "overDays": over_days,
                    "sharedMP": shared_mp,
                    "score": score,
                }
            )

        scored.sort(key=lambda s: s["score"])
        best = scored[0]

        # Update running capacity with proposed move load
        if best["mId"] in run_cap:
            for di in range(g_n_days):
                if di >= len(run_cap[best["mId"]]):
                    run_cap[best["mId"]].append({"prod": 0, "setup": 0})
                add_prod = (op.d[di] / tool.pH) * 60 if di < len(op.d) and op.d[di] > 0 else 0
                first_prod_day = next((i for i, v in enumerate(op.d) if v > 0), 0)
                add_setup = setup_min_val if di == first_prod_day or di == 0 else 0
                run_cap[best["mId"]][di]["prod"] += add_prod
                run_cap[best["mId"]][di]["setup"] += add_setup

        reasoning.append(f"Alt. disponível: {best['mId']}.")
        if best["overDays"]:
            prior_count = sum(1 for d in decs if d.action and d.action.get("toM") == best["mId"])
            reasoning.append(
                f"{best['mId']} sobrecarga {len(best['overDays'])}d "
                f"(inclui {prior_count} ops já propostas)."
            )
        else:
            reasoning.append(f"Capacidade {best['mId']}: pico {best['peak'] * 100:.0f}% — OK.")
        reasoning.append(f"Setup: +{setup_min_val:.0f}min ({tool.sH}h).")
        if best["sharedMP"]:
            reasoning.append(f"MP {tool.mp} partilhada — agrupar.")
        if tool.op > 1:
            reasoning.append(f"Requer {tool.op} operadores.")
        if not has_stk and tool.lt > 0:
            reasoning.append("STOCK ZERO — OTD em risco.")
        reasoning.append(f"→ Mover {b.tool_id} → {best['mId']}.")

        decs.append(
            ReplanProposal(
                id=f"D_{b.op_id}_RP",
                op_id=b.op_id,
                type="replan",
                severity=severity,
                title=f"{b.tool_id} → {best['mId']}",
                desc=f"{b.nm} ({b.sku})",
                reasoning=reasoning,
                impact={
                    "fromM": b.orig_m,
                    "toM": best["mId"],
                    "setupMin": setup_min_val,
                    "pcs": total_pcs,
                    "hrs": f"{total_h:.1f}",
                    "destPeak": f"{best['peak'] * 100:.0f}",
                    "overDays": len(best["overDays"]),
                    "ops": tool.op,
                    "stockRisk": not has_stk and tool.lt > 0,
                    "atr": op.atr,
                    "sharedMP": best["sharedMP"],
                    "dLoad": best["dLoad"],
                },
                action={"opId": b.op_id, "toM": best["mId"]},
            )
        )

    # Lote económico warnings
    for b in blocks:
        if b.below_min_batch and b.type == "ok":
            tool = tool_map.get(b.tool_id)
            if not tool:
                continue
            decs.append(
                ReplanProposal(
                    id=f"D_{b.op_id}_LT",
                    op_id=b.op_id,
                    type="replan",
                    severity="low",
                    title=f"{b.sku} abaixo lote econ.",
                    desc=f"{_fmt_num(b.qty)} < {_fmt_num(tool.lt)} pcs",
                    reasoning=[
                        f"Qty {_fmt_num(b.qty)} abaixo lote económico {_fmt_num(tool.lt)}.",
                        "Considerar agrupar com próxima encomenda.",
                    ],
                    impact={"qty": b.qty, "lotEconomic": tool.lt},
                    action=None,
                )
            )

    sev_order = {"critical": 0, "high": 1, "medium": 2, "low": 3}

    def _sort_decs(d: ReplanProposal) -> tuple:
        type_order = 0 if d.type == "replan" else 1
        return (type_order, sev_order.get(d.severity, 3))

    decs.sort(key=_sort_decs)
    return decs
